evm/gas-service.js

```diff
@@ -0,0 +1,414 @@
+                            decimals,
+                        },
+                        execute_gas_multiplier: multiplier = 1.1,
+                    } = result;
+
+                    const axelarBaseFee = Math.ceil(parseFloat(sourceBaseFee) * Math.pow(10, decimals));
+                    const relativeGasPrice = Math.ceil(parseFloat(gasPrice) * parseFloat(multiplier));
+
+                    return [feeType, axelarBaseFee, relativeGasPrice, blobBaseFee * relativeGasPrice];
+                });
+        }),
+    );
+}
+
+async function processCommand(_, chain, options) {
+    const {
+        env,
+        contractName,
+        address,
+        action,
+        privateKey,
+
+        txHash,
+        logIndex,
+
+        receiver,
+        token,
+        amount,
+
+        chains,
+
+        collectorReceiver,
+        collectTokens,
+        collectAmounts,
+
+        gasToken,
+        gasFeeAmount,
+        refundAddress,
+
+        destinationChain,
+        destinationAddress,
+        executionGasLimit,
+
+        yes,
+    } = options;
+
+    const contracts = chain.contracts;
+    const contractConfig = contracts[contractName];
+
+    let GasServiceAddress;
+
+    if (isValidAddress(address)) {
+        GasServiceAddress = address;
+    } else {
+        if (!contractConfig?.address) {
+            throw new Error(`Contract ${contractName} is not deployed on ${chain.name}`);
+        }
+
+        GasServiceAddress = contractConfig.address;
+    }
+
+    const rpc = chain.rpc;
+    const provider = getDefaultProvider(rpc);
+
+    const wallet = await getWallet(privateKey, provider, options);
+    const { address: walletAddress } = await printWalletInfo(wallet, options);
+
+    printInfo('Contract name', contractName);
+    printInfo('Contract address', GasServiceAddress);
+
+    const gasService = new Contract(GasServiceAddress, IAxelarGasService.abi, wallet);
+
+    const gasOptions = await getGasOptions(chain, options, contractName);
+
+    printInfo('GasService Action', action);
+    printInfo('Wallet', walletAddress);
+
+    if (prompt(`Proceed with action ${action} on chain ${chain.name}?`, yes)) {
+        return;
+    }
+
+    switch (action) {
+        case 'estimateGasFee': {
+            validateParameters({
+                isNonEmptyString: { destinationChain },
+                isValidAddress: { destinationAddress },
+                isNumber: { executionGasLimit },
+            });
+
+            const payload = options.payload || '0x';
+
+            const gasEstimate = await gasService.estimateGasFee(destinationChain, destinationAddress, payload, executionGasLimit);
+
+            printInfo('Gas Estimate', gasEstimate.toString());
+
+            break;
+        }
+
+        case 'updateGasInfo': {
+            validateParameters({
+                isNonEmptyStringArray: { chains },
+            });
+
+            const gasUpdates = await getGasInfo(env, chain, chains);
```

Comment:
If we're running this script for multiple chains `-n`, we should cache the info and reuse the same one across all chains. You can use a global variable for now. But we should introduce a state object that we pass around between `processCommands`

Reply:
It's different info for each source chain, and a different API call.

Reply:
ah, yes, we can still do it, but we'd have to normalize to the src token price ourselves by src chain. we can add it as a follow up optimization (my concern is that we might hit a rate limit due to n^2 API calls for n chains).